core/profiles/src/main/java/software/amazon/awssdk/profiles/ProfileFile.java

```diff
@@ -120,6 +129,53 @@ public Map<String, Profile> profiles() {
         return profileMap != null ? Collections.unmodifiableMap(profileMap) : profileMap;
     }
 
+    /**
+     * Checks if the profile files have been changed on disk.
+     * This method will always return {@link Boolean#FALSE} if the {@link ProfileFile} instance was built
+     * using {@link Builder#content(InputStream)} since a stream cannot be re-run
+     * or checked if it's been updated.
+     *
+     * @return True if any files used to build this object have been modified.
+     * @see #isStaleAsOf(Instant)
+     */
+    public boolean isStale() {
+        return isStaleIfAny(details -> {
+            Path contentLocation = details.getContentLocation();
+            try {
+                Instant modificationInstant = Files.getLastModifiedTime(contentLocation).toInstant();
+
+                return isStaleAsOf(modificationInstant);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        });
+    }
+
+    /**
+     * Checks if the profile files have been changed on disk as of a particular instant.
+     *
+     * @param instant The time to check against for any file modifications.
+     * @return True if any files used to build this object have been modified.
+     * @see #isStale()
+     */
+    public boolean isStaleAsOf(Instant instant) {
+        return isStaleIfAny(details -> details.wasBuiltBefore(instant));
+    }
+
+    /**
+     * Determines if there are disk modifications and creates a new instance if so.
+     *
+     * @return Either the same instance in case disk has no recent updates, or a newly constructed instance.
+     * @see #isStale()
+     */
+    public ProfileFile reload() {
```

Comment:
I think we should extract this functionality out this class. `ProfileFile` is meant to be immutable so it seems strange that it's able to do this reloading. Moving this functionality out to somewhere else will help simplify things as well since the scope of responsibility will be small

Reply:
Agreed. I've created a `ProfileFileRefresher` class and refactored some of the code into it.